### ISSUE TITLE
Repeated spaces on shebang line stops option parsing

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7416]
Migrated from [rt.perl.org#30660](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=30660) (status was 'resolved')

Searchable as RT30660$

**Comment by p5pRT** [comment]
### From tim@consultix-inc.com

#### Created by tim@consultix-inc.com



I've either found a bug in Perl\, or there's an undocumented
"feature" that I've suddenly run into for the first time\.
Either way\, /I don't like it/\! 8\-\{ 

In short\, a parsing rule \(or bug\) seems to be operating that says
after two \(or more\) spaces are seen on the shebang line\, no
additional options are recognized\! So when one tries to use \-s
for switch processing\, and that switch is preceded by two spaces\,
the result is that the \-switch=whatever argument
is /retained/ on the command line\, rather than used to set a
variable\! When the \-w option is in the ignored position\, that
fails to enable warnings\.

AFAIK\, this is not documented behavior\, and IMHO\, it is not
/desirable/ behavior\.

I've included sample programs and output below\, and confirmed that
the same results are obtained with Perl versions 5\.8\.4 and 5\.8\.2\.

\-Tim Maher
tim@&#8203;teachmeperl\.com

PROGRAMS&#8203;:

The \_sw ending on a test program means the \-s option is first\,
and the \-w second\, and vice versa\. The program invocation in each
case is "\./scriptname \-switch"\, which should set the $switch
variable in the script to 1\. Note that there's nothing
switch\-specific about this "two\-space" bug \-\- that's just the
context in which I originally found it\, and it demonstrates the
faulty behavior well\, by \(erroneously\) leaving the \-switch
argument in @&#8203;ARGV\, and triggering a "possible typo" warning
regarding $switch in that case\.

The two\_space\* versions of the programs are expected to work like
the one\_space ones\, but they don't\.

2004\-07\-05 12&#8203;:31                   one\_space\_sw                   Page 1

\#\! /usr/bin/perl \-s \-w
$switch == 1;
warn "Arguments are&#8203;: @&#8203;ARGV";


2004\-07\-05 12&#8203;:31                   one\_space\_ws                   Page 1

\#\! /usr/bin/perl \-w \-s
$switch == 1;
warn "Arguments are&#8203;: @&#8203;ARGV";


2004\-07\-05 12&#8203;:32                  two\_spaces\_sw                   Page 1

\#\! /usr/bin/perl \-s  \-w
$switch == 1;
warn "Arguments are&#8203;: @&#8203;ARGV";


2004\-07\-05 12&#8203;:32                  two\_spaces\_ws                   Page 1

\#\! /usr/bin/perl \-w  \-s
$switch == 1;
warn "Arguments are&#8203;: @&#8203;ARGV";


OUTPUT&#8203;:

RUNNING 'one\_space\_sw'&#8203;:
Useless use of numeric eq \(==\) in void context at \./one\_space\_sw line 3\.
Arguments are&#8203;:  at \./one\_space\_sw line 5\.

\(That's the correct result; warnings enabled\, and switch\-argument processed\)

RUNNING 'one\_space\_ws'&#8203;:
Useless use of numeric eq \(==\) in void context at \./one\_space\_ws line 3\.
Arguments are&#8203;:  at \./one\_space\_ws line 5\.

\(That's the correct result; warnings enabled\, and switch\-argument processed\)

RUNNING 'two\_spaces\_sw'&#8203;:
Arguments are&#8203;:  at \./two\_spaces\_sw line 5\.

\(That's the WRONG result; warnings DISabled;
&nbsp;    only switch\-option recognized\)

RUNNING 'two\_spaces\_ws'&#8203;:
Useless use of numeric eq \(==\) in void context at \./two\_spaces\_ws line 3\.
Name "main&#8203;::switch" used only once&#8203;: possible typo at \./two\_spaces\_ws line 3\.
Use of uninitialized value in numeric eq \(==\) at \./two\_spaces\_ws line 3\.
Arguments are&#8203;: \-switch at \./two\_spaces\_ws line 5\.

\(That's the WRONG result; switch not processed;
&nbsp;    only warning option recognized\)
 
==============================================================
| Tim Maher\, Ph\.D\.                    tim\(AT\)TeachMePerl\.com | 
| Seattle Perl Users Group        http&#8203;://www\.SeattlePerl\.com |
| SPUG Wiki Site               http&#8203;://Spugwiki\.Perlocity\.org |
| Perl Certification Site      http&#8203;://Perlcert\.Perlocity\.org |
==============================================================

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.2:

Configured by root at Sat Nov 15 08:54:26 PST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 2) configuration:
  Platform:
    osname=linux, osvers=2.4.19-4gb, archname=i686-linux-thread-multi
    uname='linux jumpy 2.4.19-4gb #1 mon aug 4 23:38:42 utc 2003 i686 unknown '
    config_args='-Dusethreads'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing'
    ccversion='', gccversion='3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/ucblib
    libs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.2:
    /local/perl5/site_perl/5.8.2
    /usr/lib/perl5/site_perl/5.8.2
    /local/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl/5.8.0
    /local/perl5/site_perl/5.6.0
    /local/perl5/site_perl/5.005
    /local/perl5/site_perl/i386-linux
    /usr/lib/perl5/site_perl/5.8.0/i586-linux-thread-multi
    /usr/local/lib/perl5/5.8.2/i686-linux-thread-multi
    /usr/local/lib/perl5/5.8.2
    /usr/local/lib/perl5/site_perl/5.8.2/i686-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.8.2
    /usr/local/lib/perl5/site_perl
    /local/perl5/site_perl/5.8.0 /usr/lib/perl5/site_perl/5.8.0 /local/perl5/site_perl/5.6.0 /local/perl5/site_perl/i386-linux /local/perl5/site_perl/5.005
    .


Environment for perl v5.8.2:
    HOME=/home/tim
    LANG=en_US
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/tim/bin:/usr/local/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/games:/opt/gnome2/bin:/opt/gnome/bin:/opt/kde3/bin:/opt/kde2/bin:/usr/lib/java/bin:.:/usr/X11R6/bin:/opt/kde/bin:/local/timbin:/local/dtp:/home/tim/bin:/usr/X11R6/bin:/opt/kde/bin:/local/timbin:/local/dtp:/home/tim/bin
    PERL5LIB=:/local/perl5/site_perl/5.8.2:/usr/lib/perl5/site_perl/5.8.2:/local/perl5/site_perl/5.8.0:/usr/lib/perl5/site_perl/5.8.0:/local/perl5/site_perl/5.6.0:/local/perl5/site_perl/5.005:/local/perl5/site_perl/i386-linux:/usr/lib/perl5/site_perl/5.8.0/i586-linux-thread-multi
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

Make perl recognize additional switches when more than one space is used\.\.\.

\-\- 


---
renee\.baecker@&#8203;smart\-websolutions\.de

Perl\-Magazin&#8203;: http&#8203;://foo\-magazin\.de
Perl\-Nachrichten&#8203;: http&#8203;://perl\-nachrichten\.de

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

<details><summary>perl.c.patch</summary>

```diff
--- perl.c.old	2008-08-25 10:35:38.000000000 +0200
+++ perl.c	2008-08-25 10:26:22.000000000 +0200
@@ -3370,8 +3370,10 @@
 	return s;
     case '*':
     case ' ':
-	if (s[1] == '-')	/* Additional switches on #! line. */
-	    return s+2;
+        while( *s && isSPACE(*s) )
+          ++s;
+	if (s[0] == '-')	/* Additional switches on #! line. */
+	    return s+1;
 	break;
     case '-':
     case 0:

```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, Aug 25\, 2008 at 10&#8203;:47&#8203;:31AM \+0200\, Rene Bcker wrote&#8203;:
> Make perl recognize additional switches when more than one space is used\.\.\.

Do you have an example of what this is solving? It seems a logical change to
make\, but I'm guessing as to what the problem is\.

> \-\-\- perl\.c\.old    2008\-08\-25 10&#8203;:35&#8203;:38\.000000000 \+0200
> \+\+\+ perl\.c    2008\-08\-25 10&#8203;:26&#8203;:22\.000000000 \+0200
> @&#8203;@&#8203; \-3370\,8 \+3370\,10 @&#8203;@&#8203;
>      return s;
>      case '\*'&#8203;:
>      case ' '&#8203;:
> \-    if \(s\[1\] == '\-'\)    /\* Additional switches on \#\! line\. \*/
> \-        return s\+2;
> \+        while\( \*s && isSPACE\(\*s\) \)
> \+          \+\+s;

That can be

&nbsp;       while\( isSPACE\(\*s\) \)
&nbsp;         \+\+s;

because isSPACE\('\\0'\) is going to be false\.


But is using isSPACE\(\) correct\, given that '\\r'\, '\\n' and particularly '\\t'
are handled differently in code below?
\(And no\, I'm not intimately familiar with what it is doing\, or why\)


Nicholas Clark

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

Nicholas Clark wrote&#8203;:
> On Mon\, Aug 25\, 2008 at 10&#8203;:47&#8203;:31AM \+0200\, Rene Bcker wrote&#8203;:
>   
>> Make perl recognize additional switches when more than one space is used\.\.\.
>>     
>
> Do you have an example of what this is solving? It seems a logical change to
> make\, but I'm guessing as to what the problem is\.
>   

Attached is a small script\. Without this patch it just recognizes the \-s
switch but not the \-w switch\.

>   
>> \-\-\- perl\.c\.old    2008\-08\-25 10&#8203;:35&#8203;:38\.000000000 \+0200
>> \+\+\+ perl\.c    2008\-08\-25 10&#8203;:26&#8203;:22\.000000000 \+0200
>> @&#8203;@&#8203; \-3370\,8 \+3370\,10 @&#8203;@&#8203;
>>      return s;
>>      case '\*'&#8203;:
>>      case ' '&#8203;:
>> \-    if \(s\[1\] == '\-'\)    /\* Additional switches on \#\! line\. \*/
>> \-        return s\+2;
>> \+        while\( \*s && isSPACE\(\*s\) \)
>> \+          \+\+s;
>>     
>
> That can be
>
>        while\( isSPACE\(\*s\) \)
>          \+\+s;
>
> because isSPACE\('\\0'\) is going to be false\.
>
>
> But is using isSPACE\(\) correct\, given that '\\r'\, '\\n' and particularly '\\t'
> are handled differently in code below?
> \(And no\, I'm not intimately familiar with what it is doing\, or why\)
>   

So this should be rewritten as

&nbsp;  while\( \*s && \( s\[0\] == ' ' \) \)
&nbsp;      \+\+s;

>
> Nicholas Clark
>
>
>   

Renee

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

[30660_1.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/463562/214548/30660_1.pl)

**Comment by p5pRT** [comment]
### From @tux

On Tue\, 26 Aug 2008 08&#8203;:43&#8203;:55 \+0200\, Renée Bäcker
\<renee\.baecker@&#8203;smart\-websolutions\.de> wrote&#8203;:

> Nicholas Clark wrote&#8203;:
> > On Mon\, Aug 25\, 2008 at 10&#8203;:47&#8203;:31AM \+0200\, Rene Bcker wrote&#8203;:
> >   
> >> Make perl recognize additional switches when more than one space is used\.\.\.
> >>     
> >
> > Do you have an example of what this is solving? It seems a logical change to
> > make\, but I'm guessing as to what the problem is\.
> >   
> 
> Attached is a small script\. Without this patch it just recognizes the \-s
> switch but not the \-w switch\.
> 
> >   
> >> \-\-\- perl\.c\.old    2008\-08\-25 10&#8203;:35&#8203;:38\.000000000 \+0200
> >> \+\+\+ perl\.c    2008\-08\-25 10&#8203;:26&#8203;:22\.000000000 \+0200
> >> @&#8203;@&#8203; \-3370\,8 \+3370\,10 @&#8203;@&#8203;
> >>      return s;
> >>      case '\*'&#8203;:
> >>      case ' '&#8203;:
> >> \-    if \(s\[1\] == '\-'\)    /\* Additional switches on \#\! line\. \*/
> >> \-        return s\+2;
> >> \+        while\( \*s && isSPACE\(\*s\) \)
> >> \+          \+\+s;
> >>     
> >
> > That can be
> >
> >        while\( isSPACE\(\*s\) \)
> >          \+\+s;
> >
> > because isSPACE\('\\0'\) is going to be false\.
> >
> >
> > But is using isSPACE\(\) correct\, given that '\\r'\, '\\n' and particularly '\\t'
> > are handled differently in code below?
> > \(And no\, I'm not intimately familiar with what it is doing\, or why\)
> >   
> 
> So this should be rewritten as
> 
>   while\( \*s && \( s\[0\] == ' ' \) \)
>       \+\+s;

That is just one test too many\, as it is effectively the same as

&nbsp;    while \(\*s == ' '\) s\+\+;

if \*s == \(char\)0\, \*s == ' ' will be false
If you also want to allow tabs\,

&nbsp;    while \(\*s == ' ' || \*s == '\\t'\) s\+\+;

\-\- 
H\.Merijn Brand          Amsterdam Perl Mongers  http&#8203;://amsterdam\.pm\.org/
using & porting perl 5\.6\.2\, 5\.8\.x\, 5\.10\.x\, 5\.11\.x on HP\-UX 10\.20\, 11\.00\,
11\.11\, 11\.23\, and 11\.31\, SuSE 10\.1\, 10\.2\, and 10\.3\, AIX 5\.2\, and Cygwin\.
http&#8203;://mirrors\.develooper\.com/hpux/           http&#8203;://www\.test\-smoke\.org/
http&#8203;://qa\.perl\.org      http&#8203;://www\.goldmark\.org/jeff/stupid\-disclaimers/

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Aug 26\, 2008 at 09&#8203;:16&#8203;:16AM \+0200\, H\.Merijn Brand wrote&#8203;:
> On Tue\, 26 Aug 2008 08&#8203;:43&#8203;:55 \+0200\, Renée Bäcker
> \<renee\.baecker@&#8203;smart\-websolutions\.de> wrote&#8203;:

> > > Do you have an example of what this is solving? It seems a logical change to
> > > make\, but I'm guessing as to what the problem is\.
> > >   
> > 
> > Attached is a small script\. Without this patch it just recognizes the \-s
> > switch but not the \-w switch\.

Thanks\. I used a variant in my test\.

> > > That can be
> > >
> > >        while\( isSPACE\(\*s\) \)
> > >          \+\+s;
> > >
> > > because isSPACE\('\\0'\) is going to be false\.
> > >
> > >
> > > But is using isSPACE\(\) correct\, given that '\\r'\, '\\n' and particularly '\\t'
> > > are handled differently in code below?
> > > \(And no\, I'm not intimately familiar with what it is doing\, or why\)
> > >   
> > 
> > So this should be rewritten as
> > 
> >   while\( \*s && \( s\[0\] == ' ' \) \)
> >       \+\+s;
> 
> That is just one test too many\, as it is effectively the same as
> 
>     while \(\*s == ' '\) s\+\+;
> 
> if \*s == \(char\)0\, \*s == ' ' will be false
> If you also want to allow tabs\,
> 
>     while \(\*s == ' ' || \*s == '\\t'\) s\+\+;

Tabs are handled differently in the code below

&nbsp;    case '\\n'&#8203;:
&nbsp;    case '\\t'&#8203;:
&nbsp;    break;

so it didn't seem logical to include them here\.

I applied the appended change\.

Nicholas Clark

Change 34234 by nicholas@&#8203;nicholas\-saigo on 2008/08/28 13&#8203;:11&#8203;:44

&nbsp;    Fix \#30660&#8203;: Repeated spaces on shebang line stops option parsing
&nbsp;    From a patch and test sent by Renée Bäcker in
&nbsp;    \<48B271A3\.80808@&#8203;smart\-websolutions\.de>

Affected files \.\.\.

\.\.\. //depot/perl/perl\.c\#874 edit
\.\.\. //depot/perl/t/run/switches\.t\#28 edit

Differences \.\.\.

==== //depot/perl/perl\.c\#874 \(text\) ====

@&#8203;@&#8203; \-3431\,8 \+3431\,10 @&#8203;@&#8203;
&nbsp;     return s;
&nbsp;     case '\*'&#8203;:
&nbsp;     case ' '&#8203;:
\-    if \(s\[1\] == '\-'\)    /\* Additional switches on \#\! line\. \*/
\-        return s\+2;
\+        while\( \*s == ' ' \)
\+          \+\+s;
\+    if \(s\[0\] == '\-'\)    /\* Additional switches on \#\! line\. \*/
\+        return s\+1;
&nbsp;     break;
&nbsp;     case '\-'&#8203;:
&nbsp;     case 0&#8203;:

==== //depot/perl/t/run/switches\.t\#28 \(text\) ====

@&#8203;@&#8203; \-11\,7 \+11\,7 @&#8203;@&#8203;
 
 BEGIN \{ require "\./test\.pl"; \}
 
\-plan\(tests => 69\);
\+plan\(tests => 70\);
 
 use Config;
 
@&#8203;@&#8203; \-350\,3 \+350\,19 @&#8203;@&#8203;
&nbsp;     stdin       => 'zomtek'\,
 \);
 is\( $r\, "affe\\n"\, '\-E works outside of the block created by \-n' \);
\+
\+\# RT \#30660
\+
\+$filename = tempfile\(\);
\+SKIP&#8203;: \{
\+    open my $f\, ">$filename" or skip\( "Can't write temp file $filename&#8203;: $\!" \);
\+    print $f \<\<'SWTEST';
\+\#\!perl \-w    \-iok
\+print "$^I\\n";
\+SWTEST
\+    close $f or die "Could not close&#8203;: $\!";
\+    $r = runperl\(
\+    progfile    => $filename\,
\+    \);
\+    like\( $r\, qr/ok/\, 'Spaces on the \#\! line \(\#30660\)' \);
\+\}

**Comment by p5pRT** [comment]
module@renee-baecker.de - Status changed from 'open' to 'resolved'
